README.md
# tap-holidays
[![PyPI version](https://badge.fury.io/py/tap-holidays.svg)](https://badge.fury.io/py/tap-holidays)
![PyPI - Status](https://img.shields.io/pypi/status/tap-holidays)
[![Build Status](https://travis-ci.com/goodeggs/tap-holidays.svg?branch=master)](https://travis-ci.com/goodeggs/tap-holidays.svg?branch=master)
![PyPI - Python Version](https://img.shields.io/pypi/pyversions/tap-holidays)
[![License: GPL v3](https://img.shields.io/badge/License-GPLv3-blue.svg)](https://www.gnu.org/licenses/gpl-3.0)

A [Singer](https://www.singer.io/) tap for extracting data from the [Holiday API](https://holidayapi.com/docs).

## Installation

Since package dependencies tend to conflict between various taps and targets, Singer [recommends](https://github.com/singer-io/getting-started/blob/master/docs/RUNNING_AND_DEVELOPING.md#running-singer-with-python) installing taps and targets into their own isolated virtual environments:

### Install Holidays Tap

```bash
$ python3 -m venv ~/.venvs/tap-holidays
$ source ~/.venvs/tap-holidays/bin/activate
$ pip3 install tap-holidays
$ deactivate
```

### Install Stitch Target (optional)

```bash
$ python3 -m venv ~/.venvs/target-stitch
$ source ~/.venvs/target-stitch/bin/activate
$ pip3 install target-stitch
$ deactivate
```

## Configuration

The tap accepts a JSON-formatted configuration file as arguments. This configuration file has a single required field:

1. `api_key`: A valid Holiday API key.

An bare-bones Holiday API configuration may file may look like the following:

```json
{
  "api_key": "foo"
}
```

### Granular Stream Configuration

Additionally, you may specify more granular configurations for individual streams. Each key under a stream should represent a valid API request parameter for that endpoint. A more fleshed-out configuration file may look similar to the following:

```json
{
  "api_key": "foo",
  "api_version": "v1",
  "streams": {
    "holidays": {
      "country": "US",
      "year": 2018,
      "subdivisions": true
    }
  }
}
```

## Streams

The current version of the tap syncs a single [Stream](https://github.com/singer-io/getting-started/blob/master/docs/SYNC_MODE.md#streams):
1. `Holidays`: [Endpoint Documentation](https://holidayapi.com/docs)

## Discovery

Singer taps describe the data that a stream supports via a [Discovery](https://github.com/singer-io/getting-started/blob/master/docs/DISCOVERY_MODE.md#discovery-mode) process. You can run the Dayforce tap in Discovery mode by passing the `--discover` flag at runtime:

```bash
$ ~/.venvs/tap-holidays/bin/tap-holidays --config=config/holidays.config.json --discover
```

The tap will generate a [Catalog](https://github.com/singer-io/getting-started/blob/master/docs/DISCOVERY_MODE.md#the-catalog) to stdout. To pass the Catalog to a file instead, simply redirect it to a file:s

```bash
$ ~/.venvs/tap-holidays/bin/tap-holidays --config=config/holidays.config.json --discover > catalog.json
```

## Sync to stdout

Running a tap in [Sync mode](https://github.com/singer-io/getting-started/blob/master/docs/SYNC_MODE.md#sync-mode) will extract data from the various selected Streams. In order to run a tap in Sync mode and have messages emitted to stdout, pass a valid configuration file and catalog file:

```bash
$ ~/.venvs/tap-holidays/bin/tap-holidays --config=config/holidays.config.json --catalog=catalog.json
```

The tap will emit occasional [Metric](https://github.com/singer-io/getting-started/blob/master/docs/SYNC_MODE.md#metric-messages), [Schema](https://github.com/singer-io/getting-started/blob/master/docs/SPEC.md#schema-message), [Record](https://github.com/singer-io/getting-started/blob/master/docs/SPEC.md#record-message), and [State messages](https://github.com/singer-io/getting-started/blob/master/docs/SPEC.md#state-message). You can persist State between runs by redirecting messages to a file:

```bash
$ ~/.venvs/tap-holidays/bin/tap-holidays --config=config/holidays.config.json --catalog=catalog.json >> state.json
$ tail -1 state.json > state.json.tmp
$ mv state.json.tmp state.json
```

## Sync to Stitch

You can also send the output of the tap to [Stitch Data](https://www.stitchdata.com/) for loading into the data warehouse. To do this, first create a JSON-formatted configuration for Stitch. This configuration file has two required fields:
1. `client_id`: The ID associated with the Stitch Data account you'll be sending data to.
2. `token` The token associated with the specific [Import API integration](https://www.stitchdata.com/docs/integrations/import-api/) within the Stitch Data account.

An example configuration file will look as follows:

```json
{
  "client_id": 1234,
  "token": "foobarfoobar"
}
```

Once the configuration file is created, simply pipe the output of the tap to the Stitch Data target and supply the target with the newly created configuration file:

```bash
$ ~/.venvs/tap-holidays/bin/tap-holidays --config=config/dayforce.config.json --catalog=catalog.json --state=state.json | ~/.venvs/target-stitch/bin/target-stitch --config=config/stitch.config.json >> state.json
$ tail -1 state.json > state.json.tmp
$ mv state.json.tmp state.json
```

## Contributing

The first step to contributing is getting a copy of the source code. First, [fork `tap-holidays` on GitHub](https://github.com/goodeggs/tap-holidays/fork). Then, `cd` into the directory where you want your copy of the source code to live and clone the source code:

```bash
$ git clone git@github.com:YourGitHubName/tap-holidays.git
```

Now that you have a copy of the source code on your local machine, you can leverage [Pipenv](https://docs.pipenv.org/en/latest/) and the corresponding `Pipfile` to install of the development dependencies within a virtual environment:

```bash
$ pipenv install --three --dev
```

This command will create an isolated virtual environment for your `tap-holidays` project and install all the development dependencies defined within the `Pipfile` inside of the environment. You can then enter a shell within the environment:

```bash
$ pipenv shell
```

Or, you can run individual commands within the environment without entering the shell:

```bash
$ pipenv run <command>
```

For example, to format your code using [isort](https://github.com/timothycrosley/isort) and [flake8](http://flake8.pycqa.org/en/latest/index.html) before commiting changes, run the following commands:

```bash
$ pipenv run make isort
$ pipenv run make flake8
```

You can also run the entire testing suite before committing using [tox](https://tox.readthedocs.io/en/latest/):

```bash
$ pipenv run tox
```

Finally, you can run your local version of the tap within the virtual environment using a command like the following:

```bash
$ pipenv run tap-holidays --config=config/dayforce.config.json --catalog=catalog.json
```

Once you've confirmed that your changes work and the testing suite passes, feel free to put out a PR!


CHANGELOG.md
## 0.1.1

 - Added fibonaccial decay backoff to the `_get()` method on the `HolidayAPIStream`.
 - Made Rollbar logging more explicitly optional.


tests/conftest.py
import argparse
import json

import pytest

from tap_holidays.client import HolidayStream


@pytest.fixture(scope='function')
def config(shared_datadir):
    with open(shared_datadir / 'test.config.json') as f:
        return json.load(f)


@pytest.fixture(scope='function')
def args(config, shared_datadir):
    args = argparse.Namespace()
    setattr(args, 'config', config)
    setattr(args, 'state', {})
    setattr(args, 'config_path', shared_datadir / 'test.config.json')
    return args


@pytest.fixture(scope='function', params={HolidayStream})
def client(config, args, shared_datadir, request):
    return request.param(api_key=config.get("api_key"),
                         api_version=config.get("api_version"),
                         config=config,
                         state=args.state)


tests/test_client.py
import pytest
import requests
import responses
from singer.schema import Schema

from tap_holidays.client import HolidayStream, is_fatal_code


@pytest.mark.parametrize('status_code', [400, 401, 403, 404,
                                         pytest.param(500, marks=pytest.mark.xfail),
                                         pytest.param(502, marks=pytest.mark.xfail),
                                         pytest.param(503, marks=pytest.mark.xfail),
                                         pytest.param(504, marks=pytest.mark.xfail)])
def test_is_fatal_code(status_code):
    resp = requests.models.Response()
    resp.status_code = status_code
    exc = requests.exceptions.RequestException(response=resp)
    assert is_fatal_code(exc)


@pytest.mark.parametrize('pt_stream', [HolidayStream])
def test_client_from_args_class_method(pt_stream, args, config, shared_datadir):
    stream = pt_stream.from_args(args)
    assert stream.api_key == "foobar123"
    assert stream.api_version == "v1"
    assert stream.config == config
    assert stream.state == args.state


def test_load_schema(client):
    schema = client._load_schema()
    assert isinstance(schema, dict)
    assert Schema.from_dict(schema)


def test_get_successful(client):
    with responses.RequestsMock() as rsps:
        expected = {client.tap_stream_id: [{"name": "SuperBowl"}, {"name": "Thanksgiving"}]}
        rsps.add(responses.GET, f"{client.base_url}{client.api_version}/{client.tap_stream_id}",
                 json=expected, status=200)
        resp = client._get(endpoint=f"/{client.tap_stream_id}")
        assert resp == expected


@pytest.mark.xfail()
@pytest.mark.parametrize('status_code', [400, 401, 403, 404])
def test_get_fatal(client, status_code):
    with responses.RequestsMock() as rsps:
        expected = {}
        rsps.add(responses.GET, f"{client.base_url}{client.api_version}/{client.tap_stream_id}",
                 json=expected, status=status_code)
        resp = client._get(endpoint=f"/{client.tap_stream_id}")
        assert resp == expected


tap_holidays/__init__.py
import json
import os

import rollbar
import singer

from .client import HolidayStream

LOGGER = singer.get_logger()

try:
    ROLLBAR_ACCESS_TOKEN = os.environ["ROLLBAR_ACCESS_TOKEN"]
    ROLLBAR_ENVIRONMENT = os.environ["ROLLBAR_ENVIRONMENT"]
except KeyError:
    LOGGER.info("No Rollbar environment variables found. Rollbar logging disabled..")
    log_to_rollbar = False
else:
    rollbar.init(ROLLBAR_ACCESS_TOKEN, ROLLBAR_ENVIRONMENT)
    log_to_rollbar = True


REQUIRED_CONFIG_KEYS = [
    "api_key"
]

AVAILABLE_STREAMS = {
    HolidayStream
}


def discover(args, select_all=False):
    LOGGER.info('Starting discovery..')

    catalog = {"streams": []}
    for available_stream in AVAILABLE_STREAMS:
        stream = available_stream.from_args(args)
        catalog_entry = {
            "stream": stream.tap_stream_id,
            "tap_stream_id": stream.tap_stream_id,
            "schema": stream.schema,
            "metadata": singer.metadata.get_standard_metadata(schema=stream.schema,
                                                              key_properties=stream.key_properties,
                                                              valid_replication_keys=stream.bookmark_properties,
                                                              replication_method=stream.replication_method)
        }

        if select_all is True:
            catalog_entry["metadata"][0]["metadata"]["selected"] = True
        catalog["streams"].append(catalog_entry)

    print(json.dumps(catalog, indent=2))


def sync(args):
    LOGGER.info('Starting sync..')

    if not args.state:
        args.state = {"bookmarks": {}}

    selected_streams = {catalog_entry.stream for catalog_entry in args.catalog.get_selected_streams(args.state)}

    for available_stream in AVAILABLE_STREAMS:
        if available_stream.tap_stream_id in selected_streams:
            stream = available_stream.from_args(args)
            LOGGER.info(f"Starting sync for stream {stream.tap_stream_id}..")
            singer.bookmarks.set_currently_syncing(state=args.state, tap_stream_id=stream.tap_stream_id)
            stream.write_state_message()
            stream.write_schema_message()
            stream.sync()
            singer.bookmarks.set_currently_syncing(state=args.state, tap_stream_id=None)
            stream.write_state_message()


def main():
    args = singer.parse_args(required_config_keys=REQUIRED_CONFIG_KEYS)
    if args.discover:
        try:
            discover(args, select_all=True)
        except:
            LOGGER.exception("Caught an exception during Discovery..")
            if log_to_rollbar is True:
                rollbar.report_exc_info()
    else:
        try:
            sync(args)
        except:
            LOGGER.exception("Caught an exception during Sync..")
            if log_to_rollbar is True:
                rollbar.report_exc_info()


if __name__ == "__main__":
    main()


tap_holidays/client.py
import os
from typing import ClassVar, Dict, Iterator, List, Optional, Set

import attr
import backoff
import requests
import singer

from .version import __version__

LOGGER = singer.get_logger()


def is_fatal_code(e: requests.exceptions.RequestException) -> bool:
    '''Helper function to determine if a Requests reponse status code
    is a "fatal" status code. If it is, the backoff decorator will giveup
    instead of attemtping to backoff.'''
    return 400 <= e.response.status_code < 500 and e.response.status_code != 429


@attr.s
class HolidayAPIStream(object):
    tap_stream_id: ClassVar[Optional[str]] = None

    api_key: str = attr.ib()
    config: Dict = attr.ib(repr=False)
    state: Dict = attr.ib()
    api_version: str = attr.ib(default="v1", validator=attr.validators.instance_of(str))
    params: Dict = attr.ib(default=None)
    base_url: str = attr.ib(default="https://holidayapi.com/")

    def __attrs_post_init__(self):
        if self.tap_stream_id is not None:
            self.schema = self._load_schema()

        if self.config.get("streams") is not None:
            self.params = self.config.get("streams", {}).get(self.tap_stream_id, {})
            if not isinstance(self.params, dict):
                raise TypeError("Stream parameters must be supplied as JSON.")
            else:
                for key in self.params.keys():
                    if key not in self.valid_params:
                        raise ValueError(f"{key} is not a valid parameter for stream {self.tap_stream_id}")

                for param in self.required_params:
                    if param not in self.params.keys():
                        raise ValueError(f"{param} is a required parameter for stream {self.tap_stream_id}")

    @classmethod
    def from_args(cls, args):
        return cls(api_key=args.config.get("api_key"),
                   api_version=args.config.get("api_version"),
                   config=args.config,
                   state=args.state)

    def _get_abs_path(self, path: str) -> str:
        return os.path.join(os.path.dirname(os.path.realpath(__file__)), path)

    def _load_schema(self) -> Dict:
        '''Loads a JSON schema file for a given
        Dayforce resource into a dict representation.
        '''
        schema_path = self._get_abs_path("schemas")
        return singer.utils.load_json(f"{schema_path}/{self.tap_stream_id}.json")

    def _construct_headers(self) -> Dict:
        '''Constructs a standard set of headers for HTTPS requests.'''
        headers = requests.utils.default_headers()
        headers["User-Agent"] = f"python-holidays-tap/{__version__}"
        headers["Date"] = singer.utils.strftime(singer.utils.now(), '%a, %d %b %Y %H:%M:%S %Z')
        return headers

    @backoff.on_exception(backoff.fibo,
                          requests.exceptions.HTTPError,
                          max_time=120,
                          giveup=is_fatal_code,
                          logger=LOGGER)
    @backoff.on_exception(backoff.fibo,
                          (requests.exceptions.ConnectionError,
                           requests.exceptions.Timeout),
                          max_time=120,
                          logger=LOGGER)
    def _get(self, endpoint: str, params: Dict = None) -> Dict:
        '''Constructs a standard way of making
        a GET request to the Holiday API.
        '''
        url = self.base_url + self.api_version + endpoint
        headers = self._construct_headers()
        self.params.update({"key": self.api_key})
        response = requests.get(url, headers=headers, params=params)
        response.raise_for_status()
        return response.json()

    def _yield_records(self, entity: str, params: Dict = None) -> Iterator[Dict]:
        '''Yeild individual records for a given entity.'''
        records = self._get(endpoint=f"/{entity}", params=params).get(entity)
        for record in records:
            yield record

    def sync(self):
        '''Sync data according to Singer spec.'''
        with singer.metrics.job_timer(job_type=f"sync_{self.tap_stream_id}"):
            with singer.metrics.record_counter(endpoint=self.tap_stream_id) as counter:
                for record in self._yield_records(entity=self.tap_stream_id, params=self.params):
                    with singer.Transformer() as transformer:
                        transformed_record = transformer.transform(data=record, schema=self.schema)
                        singer.write_record(stream_name=self.tap_stream_id, time_extracted=singer.utils.now(), record=transformed_record)
                        counter.increment()

    def write_schema_message(self):
        '''Writes a Singer schema message.'''
        return singer.write_schema(stream_name=self.tap_stream_id, schema=self.schema, key_properties=self.key_properties)

    def write_state_message(self):
        '''Writes a Singer state message.'''
        return singer.write_state(self.state)


@attr.s
class HolidayStream(HolidayAPIStream):
    tap_stream_id: ClassVar[str] = 'holidays'
    key_properties: ClassVar[List[str]] = ["uuid", "date"]
    bookmark_properties: ClassVar[List[str]] = []
    replication_method: ClassVar[str] = 'full_table'
    required_params: ClassVar[Set[str]] = {
        "country",
        "year"
    }
    valid_params: ClassVar[Set[str]] = {
        "country",
        "year",
        "month",
        "day",
        "previous",
        "upcoming",
        "public",
        "subdivisions",
        "search",
        "language"
    }
